GraphicsLayerWC: site isolation support

fujii opened the pull request:
#### dc5a423def84441a34100343e210d96406307e04
<pre>
GraphicsLayerWC: site isolation support
<a href="https://bugs.webkit.org/show_bug.cgi?id=254902">https://bugs.webkit.org/show_bug.cgi?id=254902</a>

Reviewed by Don Olmstead.

This is the initial support of site isolation for GraphicsLayerWC.
Cocoa port (aka RemoteLayerTree and GraphicsLayerCARemote) transfers a
layer tree to UI process, GraphicsLayerWC transfers a layer tree to
GPU process. GPU process renders a remote child frame into a texture,
and attach the texture into the content of iframe layer of the parent
frame&apos;s layer tree. Added WCRemoteFrameHostLayerManager class to
manage a HashMap from a LayerHostingContextIdentifier to a
BitmapTexture.

This has a problem now. Updating remote child frame&apos;s layer tree
doesn&apos;t trigger the parent frame&apos;s repaint. For the workaround, you
need to manually click a browser window of MiniBrowser to repaint.
Keep http/tests/site-isolation tests skipped for WinCairo.

MiniBrowser has to enable SiteIsolationEnabled option to test it by
using the internal settings menu item or adding the following code.
&gt; WKPreferencesSetBoolValueForKeyForTesting(preferences.get(), true, createWKString(&quot;SiteIsolationEnabled&quot;).get());

Fixed a bug of TextureMapperGL::endPainting. It should restore a
previous frame buffer that TextureMapperGL::beginPainting saved.
TextureMapperGL supports the offscreen rendering mode for
WebKitTestRunner pixel dump mode. It&apos;s not a problem for
WebKitTestRunner because it renders a layer tree always to a texture.
In site-isolation, however, remote frame&apos;s layer tree renders to a
texture and main frame layer tree to a frame buffer. After rendering
to a texture, it has to restore the previous frame buffer.

* Source/WebCore/platform/graphics/texmap/TextureMapperGL.cpp:
(WebCore::TextureMapperGL::endPainting): Restore the previous framebuffer.
* Source/WebKit/GPUProcess/GPUConnectionToWebProcess.cpp:
(WebKit::GPUConnectionToWebProcess::didClose):
* Source/WebKit/GPUProcess/graphics/wc/RemoteWCLayerTreeHost.cpp:
(WebKit::remoteGraphicsStreamWorkQueue):
* Source/WebKit/GPUProcess/graphics/wc/RemoteWCLayerTreeHost.h:
* Source/WebKit/GPUProcess/graphics/wc/WCRemoteFrameHostLayerManager.cpp: Added.
(WebKit::WCRemoteFrameHostLayerManager::singleton):
(WebKit::WCRemoteFrameHostLayerManager::acquireRemoteFrameHostLayer):
(WebKit::WCRemoteFrameHostLayerManager::releaseRemoteFrameHostLayer):
(WebKit::WCRemoteFrameHostLayerManager::updateTexture):
(WebKit::WCRemoteFrameHostLayerManager::removeAllLayersForProcess):
* Source/WebKit/GPUProcess/graphics/wc/WCRemoteFrameHostLayerManager.h: Added.
* Source/WebKit/GPUProcess/graphics/wc/WCScene.cpp:
(WebKit::WCScene::update):
* Source/WebKit/PlatformPlayStation.cmake:
* Source/WebKit/PlatformWin.cmake:
* Source/WebKit/WebProcess/WebPage/wc/DrawingAreaWC.cpp:
(WebKit::DrawingAreaWC::DrawingAreaWC):
(WebKit::DrawingAreaWC::updateRootLayers):
(WebKit::DrawingAreaWC::setRootCompositingLayer):
(WebKit::DrawingAreaWC::addRootFrame):
(WebKit::DrawingAreaWC::attachViewOverlayGraphicsLayer):
(WebKit::DrawingAreaWC::updatePreferences):
(WebKit::DrawingAreaWC::shouldUseTiledBackingForFrameView const):
(WebKit::DrawingAreaWC::updateGeometryWC):
(WebKit::DrawingAreaWC::isCompositingMode):
(WebKit::DrawingAreaWC::sendUpdateAC):
(WebKit::DrawingAreaWC::rootLayerInfoWithFrameIdentifier):
* Source/WebKit/WebProcess/WebPage/wc/DrawingAreaWC.h:
* Source/WebKit/WebProcess/WebPage/wc/GraphicsLayerWC.cpp:
(WebKit::GraphicsLayerWC::setContentsToPlatformLayerHost):
(WebKit::GraphicsLayerWC::flushCompositingStateForThisLayerOnly):
* Source/WebKit/WebProcess/WebPage/wc/GraphicsLayerWC.h:
* Source/WebKit/WebProcess/WebPage/wc/WCUpdateInfo.h:
(WebKit::WCLayerUpdateInfo::encode const):
(WebKit::WCLayerUpdateInfo::decode):
(WebKit::WCUpdateInfo::encode const):
(WebKit::WCUpdateInfo::decode):

Canonical link: <a href="https://commits.webkit.org/265099@main">https://commits.webkit.org/265099@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f90ac071f40c8e8491d58b554725329730e28328

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/9834 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/10082 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/10327 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/11486 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/9558 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/9843 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/12066 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/10032 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/12491 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/9988 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/10792 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/8337 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/11891 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/8099 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/8920 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/2/builds/11891 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/9197 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/9069 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/2/builds/11891 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/9569 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/7787 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/8735 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/12959 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/1114 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/9344 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebKit/GPUProcess/graphics/wc/WCRemoteFrameHostLayerManager.cpp:
```diff
@@ -0,0 +1,111 @@
+    {
+    }
+
+    WebCore::ProcessIdentifier webProcessIdentifier() { return m_webProcessIdentifier; }
```
const

fujii replied:
Fixed

achristensen07 commented on Source/WebKit/GPUProcess/graphics/wc/WCRemoteFrameHostLayerManager.cpp:
```diff
@@ -0,0 +1,111 @@
+    WCRemoteFrameHostLayer m_layer;
+};
+
+WCRemoteFrameHostLayerManager& WCRemoteFrameHostLayerManager::singleton()
```
Does this need to be a singleton?  I think it would be better organized if it were owned by something, but I'm not sure what.

fujii replied:
GPUConnectionToWebProcess can be a owner but it resides in the main thread while WCRemoteFrameHostLayerManager resides in the OpenGL thread and is thread agnostic.
It will be a bit tricky code.

achristensen07 commented on Source/WebKit/WebProcess/WebPage/wc/DrawingAreaWC.cpp:
```diff
@@ -209,7 +227,7 @@ static void flushLayerImageBuffers(WCUpateInfo& info)
 bool DrawingAreaWC::isCompositingMode()
 {
-    return m_contentLayer;
+    return m_rootLayers[0].contentLayer;
```
We need to be working towards having no preference for the first root layer.

fujii replied:
Fixed.

achristensen07 commented on Source/WebKit/WebProcess/WebPage/wc/DrawingAreaWC.cpp:
```diff
@@ -244,32 +262,50 @@ void DrawingAreaWC::updateRendering()
+            rootLayer.viewOverlayRootLayer->flushCompositingState(visibleRect);
+
+        m_commitQueue->dispatch([this, weakThis = WeakPtr(*this), stateID = m_backingStoreStateID, updateInfo = std::exchange(m_updateInfo, { }), willCallDisplayDidRefresh]() mutable {
+            flushLayerImageBuffers(updateInfo);
```
We should probably check weakThis before using this.

fujii replied:
Fixed.

fujii replied:
WeakPtr isn't thread safe. I should nothing for the WeakPtr in another thread.

fujii replied:
Reverted the fix.

achristensen07 commented on Source/WebKit/WebProcess/WebPage/wc/WCUpateInfo.h:
```diff
@@ -308,6 +318,8 @@ struct WCUpateInfo {
     template <class Decoder>
     static WARN_UNUSED_RETURN bool decode(Decoder& decoder, WCUpateInfo& result)
     {
+        if (!decoder.decode(result.remoteContextHostedIdentifier))
```
It would be nice if all this could be converted to WCUpdateInfo.serialization.in

fujii replied:
Fixed.

achristensen07 commented on Source/WebCore/platform/graphics/texmap/TextureMapperGL.cpp:
```diff
@@ -222,6 +222,7 @@ void TextureMapperGL::beginPainting(PaintFlags flags, BitmapTexture* surface)
 
 void TextureMapperGL::endPainting()
 {
+    glBindFramebuffer(GL_FRAMEBUFFER, data().targetFrameBuffer);
```
I can't say I understand why this is needed.

fujii replied:
Added more commit message.

achristensen07 commented on Source/WebKit/GPUProcess/graphics/wc/WCRemoteFrameHostLayerManager.cpp:
```diff
@@ -0,0 +1,111 @@
+    m_layers.remove(layerHostingContextIdentifier);
+}
+
+void WCRemoteFrameHostLayerManager::updateTexture(WebCore::LayerHostingContextIdentifier layerHostingContextIdentifier, WebCore::ProcessIdentifier webProcessIdentifier, RefPtr<WebCore::BitmapTexture> texture)
```
RefPtr<WebCore::BitmapTexture>&& might reduce some ref churn.

fujii replied:
I use `WTFMove` both inside and callsite of updateTexture. There should be no ref-churn here.

achristensen07 reviewed (commented):
Thank you for being engaged in this development.  It is good to know that site isolation won't be just a Cocoa-specific development, which is important for the health of the project.

Do you know what it would take to get it so that we could unskip the test and have the drawing driven in all processes?  That would probably be worth finding out.  Also, the test times out occasionally on macOS for an unknown reason, and I am hoping to fix that soonish.

For a currently unknown reason, on macOS one isolated iframe draws fine but the second isolated iframe does not draw.  I was hoping to figure that out and fix it before cementing the architecture in even more by adding another port's implementation.

One thing I had to make sure I did right is attach the layer and sublayer correctly if the parent layer was created first and if the child layer was created first, which required two places where we stitched layers together.  I'm not sure I see that in your code.

fujii committed:
GraphicsLayerWC: site isolation support
https://bugs.webkit.org/show_bug.cgi?id=254902

Reviewed by Don Olmstead.

This is the initial support of site isolation for GraphicsLayerWC.
Cocoa port (aka RemoteLayerTree and GraphicsLayerCARemote) transfers a
layer tree to UI process, GraphicsLayerWC transfers a layer tree to
GPU process. GPU process renders a remote child frame into a texture,
and attach the texture into the content of iframe layer of the parent
frame's layer tree. Added WCRemoteFrameHostLayerManager class to
manage a HashMap from a LayerHostingContextIdentifier to a
BitmapTexture.

This has a problem now. Updating remote child frame's layer tree
doesn't trigger the parent frame's repaint. For the workaround, you
need to manually click a browser window of MiniBrowser to repaint.
Keep http/tests/site-isolation tests skipped for WinCairo.

MiniBrowser has to enable SiteIsolationEnabled option to test it by
using the internal settings menu item or adding the following code.
> WKPreferencesSetBoolValueForKeyForTesting(preferences.get(), true, createWKString("SiteIsolationEnabled").get());

Fixed a bug of TextureMapperGL::endPainting. It should restore a
previous frame buffer that TextureMapperGL::beginPainting saved.
TextureMapperGL supports the offscreen rendering mode for
WebKitTestRunner pixel dump mode. It's not a problem for
WebKitTestRunner because it renders a layer tree always to a texture.
In site-isolation, however, remote frame's layer tree renders to a
texture and main frame layer tree to a frame buffer. After rendering
to a texture, it has to restore the previous frame buffer.

* Source/WebCore/platform/graphics/texmap/TextureMapperGL.cpp:
(WebCore::TextureMapperGL::endPainting): Restore the previous framebuffer.
* Source/WebKit/GPUProcess/GPUConnectionToWebProcess.cpp:
(WebKit::GPUConnectionToWebProcess::didClose):
* Source/WebKit/GPUProcess/graphics/wc/RemoteWCLayerTreeHost.cpp:
(WebKit::remoteGraphicsStreamWorkQueue):
* Source/WebKit/GPUProcess/graphics/wc/RemoteWCLayerTreeHost.h:
* Source/WebKit/GPUProcess/graphics/wc/WCRemoteFrameHostLayerManager.cpp: Added.
(WebKit::WCRemoteFrameHostLayerManager::singleton):
(WebKit::WCRemoteFrameHostLayerManager::acquireRemoteFrameHostLayer):
(WebKit::WCRemoteFrameHostLayerManager::releaseRemoteFrameHostLayer):
(WebKit::WCRemoteFrameHostLayerManager::updateTexture):
(WebKit::WCRemoteFrameHostLayerManager::removeAllLayersForProcess):
* Source/WebKit/GPUProcess/graphics/wc/WCRemoteFrameHostLayerManager.h: Added.
* Source/WebKit/GPUProcess/graphics/wc/WCScene.cpp:
(WebKit::WCScene::update):
* Source/WebKit/PlatformPlayStation.cmake:
* Source/WebKit/PlatformWin.cmake:
* Source/WebKit/WebProcess/WebPage/wc/DrawingAreaWC.cpp:
(WebKit::DrawingAreaWC::DrawingAreaWC):
(WebKit::DrawingAreaWC::updateRootLayers):
(WebKit::DrawingAreaWC::setRootCompositingLayer):
(WebKit::DrawingAreaWC::addRootFrame):
(WebKit::DrawingAreaWC::attachViewOverlayGraphicsLayer):
(WebKit::DrawingAreaWC::updatePreferences):
(WebKit::DrawingAreaWC::shouldUseTiledBackingForFrameView const):
(WebKit::DrawingAreaWC::updateGeometryWC):
(WebKit::DrawingAreaWC::isCompositingMode):
(WebKit::DrawingAreaWC::sendUpdateAC):
(WebKit::DrawingAreaWC::rootLayerInfoWithFrameIdentifier):
* Source/WebKit/WebProcess/WebPage/wc/DrawingAreaWC.h:
* Source/WebKit/WebProcess/WebPage/wc/GraphicsLayerWC.cpp:
(WebKit::GraphicsLayerWC::setContentsToPlatformLayerHost):
(WebKit::GraphicsLayerWC::flushCompositingStateForThisLayerOnly):
* Source/WebKit/WebProcess/WebPage/wc/GraphicsLayerWC.h:
* Source/WebKit/WebProcess/WebPage/wc/WCUpdateInfo.h:
(WebKit::WCLayerUpdateInfo::encode const):
(WebKit::WCLayerUpdateInfo::decode):
(WebKit::WCUpdateInfo::encode const):
(WebKit::WCUpdateInfo::decode):

Canonical link: https://commits.webkit.org/265099@main
